scipy/sparse/_coo.py

```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
+    def _add_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._add_sparse(self, other)
 
-        if self.ndim == 2:
-            col = self.col
-            row = self.row
-        elif self.ndim == 1:
-            col = self.coords[0]
-            row = np.zeros_like(col)
-        else:
-            raise NotImplementedError(
-                f"coo_matvec not implemented for ndim={self.ndim}")
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = self.__class__((new_data, new_coords), shape=self.shape)         
+        return A
+
+    
+    def _sub_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._sub_sparse(self, other)
+
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, -other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = coo_array((new_data, new_coords), shape=self.shape)
+        return A
+    
+
+    def _matmul_vector(self, other):
+        if self.ndim < 3:
+            result_shape = self.shape[0] if self.ndim > 1 else 1
+            result = np.zeros(result_shape,
+                              dtype=upcast_char(self.dtype.char, other.dtype.char))
+            if self.ndim == 2:
+                col = self.col
+                row = self.row
+            elif self.ndim == 1:
+                col = self.coords[0]
+                row = np.zeros_like(col)
+            else:
+                raise NotImplementedError(
+                    f"coo_matvec not implemented for ndim={self.ndim}")
+
+            coo_matvec(self.nnz, row, col, self.data, other, result)
+            # Array semantics return a scalar here, not a single-element array.
+            if isinstance(self, sparray) and result_shape == 1:
+                return result[0]
+            return result
+        
+        else: # if dim >= 3
+            result = np.zeros(math.prod(self.shape[:-1]),
+                              dtype=upcast_char(self.dtype.char, other.dtype.char))
+            shape = np.array(self.shape)
+            coords = np.concatenate(self.coords)
+            coo_matvec_nd(self.nnz, len(self.shape), shape, coords, self.data,
+                          other, result)
+            
+            result = result.reshape(self.shape[:-1])
+            return result
+
+
+    def _matmul_dispatch(self, other):
+        if self.ndim < 3:
+            return _data_matrix._matmul_dispatch(self, other)
+        
+        N = self.shape[-1]
+        if other.shape == (N,):
```

Comment:
We need to check that `other` isn't sparse before dispatching here, because 3d-sparse @ 1d-sparse won't work in this code path.

Reply:
Ah yes, I missed that out. Also, I noticed we called `_spbase._matmul_dispatch` if `self.ndim<3`, irrespective of `other`'s dimension. This does throw an error at some point if we multiply, example `2-D * 5-D`, because we don't have n-D `matmul` yet, but it will eventually cause issues when we add `matmul` in the next PR, so I just extended the conditions checked by the if statement. Do the fixes at e6004d3ad2a46dc00ebc38ad40791a6750ae7865 work fine?

Reply:
I think we want to avoid the computation used here to figure out whether to use the existing 2D matmul, especially the `asarray(other)` that is only used to get its shape. 
Furthermore, I think you shouldn't try to cover the case of `other` being sparse with large ndim here. The `_spbase._matmul_dispatch` call will return `NotImplemented` and Python will try `__rmatmul__` which will call `_rmatmul_dispatch` which switches the arguments using transposes and then calls this with the shapes changed.   So, I think the initial check can just be 
```python
if self.ndim < 3:
    return _spbase._matmul(self, other)
```

I think I don't understand the approach you are using for the rest of the `_matmul_dispatch` code here. It looks very similar to the 2D version but stops before handling any array-like (usually a list-of-lists). Can you briefly say here how this differs from the 2D approach? for example, are we removing the ability to matmul with a 2D list of lists?

Reply:
I'll remove the `asarray(other)` computation and revert the if condition back to what it was.

As of the array-like cases, I originally planned on handling that altogether when all n-D cases of matmul were working, but I think it may be wiser to have matrix-vector multiplication fully working in this PR, so I will incorporate those cases too.

While exploring the current 2-D mat-vec matmul, I noticed the following variations in behaviour -
```py
>>> a = coo_array([1,2,3])
>>> b = coo_array([[4],[5],[6]], shape=(3,1))
>>> dense = a.toarray() @ b.toarray()
>>> sparse_dense = a @ b.toarray()
>>> sparse_denselist = a @ list(b.toarray())
>>> sparse_sparse = (a@b).toarray()
>>> dense
array([32])
>>> sparse_sparse
array([32])
>>> sparse_dense
np.int64(32)
>>> sparse_denselist
array([[32]])
```

Was this intentional, or is this something I should fix for existing 2-D matmul, so that it is consistent with `np.matmul`'s behaviour?

Reply:
The rules we intend to support for matmul are the [array_api standards for matmul](https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.matmul.html). There are two descriptions on that page (one about handling input shapes and the other handling output shapes).

It might make sense to split changes to existing 1/2D code into a separate PR, if only to keep them small. But if that makes it more complicated to review than a single PR go with a single PR. Review time seems to be the restricting features of code maintenance and upkeep.

One difficulty with this part of the code is that it was originally written for 2D systems so "vec" meant a 2D row or column. It is now tempting to think of a 1D object as a "vec" -- and that often works, but not always. The best approach of course is to do what you did in this comment and ignore the method names -- instead try the top level expressions to see if they mimic what numpy/array_api makes happen.

Reply:
I've tested the top level expressions and they pass the matmul tests against numpy's behaviour now. If you'd prefer, I can make a separate PR for the 1/2D code change.
The only issue with that might be that it would take time to be reviewed, hence blocking this conversation from being resolved.
It's only the following if statement added, so if it's easier to review it in a commit in this PR itself, that also works -

Instead of
```py
if other.ndim == 1 or other.ndim == 2 and other.shape[1] == 1:
   ...
   result = self._matmul_vector(np.ravel(other))
   ...
   if other.ndim == 2 and other.shape[1] == 1:
       # If 'other' was an (nx1) column vector, reshape the result
       result = result.reshape(-1, 1)
    return result
```

we would have
```py
if other.ndim == 1 or other.ndim == 2 and other.shape[1] == 1:
    ...
    result = self._matmul_vector(np.ravel(other))
    ...
    if other.ndim == 2 and other.shape[1] == 1:
        # If 'other' was an (nx1) column vector, reshape the result
        if self.ndim == 1:
            result = result.reshape(-1)
        else:
            result = result.reshape(-1, 1)
    return result

```

(and a similar one liner in another part of the method).

Reply:
I think a PR with those two one-liners and tests that show the undesired behavior has been corrected will be quickly reviewed and merged.  I don't think that would block this conversation from being resolved... It's a separate topic from needing to check whether `other` is sparse before dispatching. I guess there could be conflicts created by one PR when the other PR is merged, but they should be easy to resolve -- there are only two lines that could conflict.

I also don't think resolving the current 1-2D shape issues will stop this PR from being reviewed/merged. That is an existing problem, not a new feature.

Note that these are only my suggestions/perspectives based on my experience. They are not factual statements. You can put the one-liners in where you think they best fit. They will fix current bugs.